### PR TITLE
Averagebutton

### PR DESCRIPTION
**Pull request by dwmoreau**
Hook up the average button in the xfel gui.

### CODE REVIEW

**Review by dwpaley — CHANGES_REQUESTED** [review]
Added some minor comments but generally looks good. I'd like to experiment with it a little before approving. Could you write down some notes on how you've been testing this?

File: xfel/command_line/cxi_mpi_submit.py [hunk 1]
```diff
@@ -218,18 +221,19 @@ def get_trialdir(output_dir, run_num, trial = None, rungroup = None, task = None
...
+          for trial_direntry in os.scandir(direntry.path):
+            if direntry.is_dir(follow_symlinks=False):
+              trial = max([
+                int(trial_direntry.name.split('_')[0]),
+                trial
+              ])
+      trial += 1
+      trialdir = os.path.join(rundir, "%03d" % trial)
+      if rungroup is not None:
+        trialdir += "_rg%03d" % rungroup
```
[the remark this answers is not in the document]
Comment (dwmoreau) [reply xfel/command_line/cxi_mpi_submit.py]: This is not needed and was removed

File: xfel/command_line/cxi_mpi_submit.py [hunk 2]
```diff
@@ -454,18 +457,22 @@ def run(self, argv = None):
...
+    if params.output.add_output_dir_option:
+      command = "%s %s output.output_dir=%s %s %s" % (
+        params.input.dispatcher, data_str, output_dir,
+        logging_str, extra_str
+      )
+    else:
+      command = "%s %s %s %s" % (
+        params.input.dispatcher, data_str,
+        logging_str, extra_str
+      )
```
[the remark this answers is not in the document]
Comment (dwmoreau) [reply xfel/command_line/cxi_mpi_submit.py]: Updated

File: xfel/ui/db/job.py [hunk 3]
```diff
@@ -103,6 +103,85 @@ def get_identifier_string(self):
...
+        self.args.append(arg)
+
+    if self.app.params.facility.name == 'lcls':
+      locator_path = os.path.join(configs_dir, identifier_string + ".loc")
+      self.args.append('input.locator = %s' % str(locator_path))
+      locator = open(locator_path, 'w')
+      locator.write("experiment=%s\n"%self.app.params.facility.lcls.experiment) # LCLS specific parameter
+      locator.write("run=%s\n"%self.run.run)
+      locator.write("detector_address=%s\n"%self.address)
+      locator.close()
```
[the remark this answers is not in the document]
Comment (dwmoreau) [reply xfel/ui/db/job.py]: Updated

File: xfel/util/mp.py [hunk 4]
```diff
@@ -512,8 +523,16 @@ def eval_params(self):
     if 'phenix' in self.command:
       self.source_env_scripts.append("cd %s\n"%os.path.dirname(self.submit_path))
 
+    if '<output_dir>' in self.command:
+      self.command = self.command.replace(
+        '<output_dir>',
+        '/'.join(self.stdoutdir.split('/')[:-1])
+      )
     # <args> (optional, following the command)
+    image_average_output_dir = '/'.join(self.stdoutdir.split('/')[:-1]) + '/out'
```
[the remark this answers is not in the document]
Comment (dwmoreau) [reply xfel/util/mp.py]: Updated in all relevant locations

File: xfel/ui/db/job.py [hunk 5]
```diff
@@ -103,6 +103,85 @@ def get_identifier_string(self):
...
+      'output.add_output_dir_option = False',
+      'mp.add_mpi_option = False',
+      'mp.extra_args = %s' % extra_args,
+      'mp.method = %s' % self.app.params.mp.method,
+    ]
+
+    if self.app.params.mp.method != 'local' or (self.app.params.mp.method == 'local' and self.app.params.facility.name == 'lcls'):
+      mp_args = [
+        'mp.use_mpi =  %s' % self.app.params.mp.use_mpi,
+        'mp.mpi_command =  %s' % self.app.params.mp.mpi_command,
```
[the remark this answers is not in the document]
Comment (dwmoreau) [reply xfel/ui/db/job.py]: Updated code in all relevant locations

File: xfel/ui/db/job.py [hunk 6]
```diff
@@ -103,6 +103,85 @@ def get_identifier_string(self):
...
+    # Make an argument list that can be submitted to cxi_mpi_submit.
+    # dxtbx.image_average does not use phil files.
+    extra_args = "-a <output_dir>/avg.cbf -m <output_dir>/max.cbf -s <output_dir>/std.cbf --mpi=True"
+    if self.skip_images > 0:
+      extra_args += ' --skip-images=%d' % self.skip_images
+    if self.num_images > 0:
+      extra_args += ' --num-images=%d' % self.num_images
+    self.args = [
+      'input.run_num = %d' % int(self.run.run),
+      'input.trial = %d' % int(self.trial.trial),
```
[the remark this answers is not in the document]
Comment (dwmoreau) [reply xfel/ui/db/job.py]: Removed casting to int

File: xfel/ui/components/averaging.py [hunk 7]
```diff
@@ -1,60 +1,14 @@
...
+def AveragingCommand(run, params, trial_num, skip_images, num_images, address, raw):
+  job = AveragingJob(run.app)
+  job.run = run
+  if run.app.params.facility.name == 'lcls':
+    job.address = address
+  job.trial = job.run.app.create_trial(trial=trial_num)
+  job.skip_images = skip_images
+  job.num_images = num_images
+  job.submit()
+  print('DONE')
```
[the remark this answers is not in the document]
Comment (dwmoreau) [reply xfel/ui/components/averaging.py]: Deleted - I believe this was used for debugging

File: xfel/ui/components/xfel_gui_dialogs.py [hunk 8]
```diff
@@ -1406,22 +1430,83 @@ def __init__(self, parent, run, params,
...
     from xfel.ui.components.averaging import AveragingCommand
-    from libtbx import easy_run
     raw = self.raw_toggle.raw.GetValue() == 1
-    average_command = AveragingCommand(self.run, self.params, raw)()
-    print("executing", average_command)
-    result = easy_run.fully_buffered(command=average_command)
-    result.show_stdout()
-    e.Skip()
+    skip_images = self.skip_images.ctr.GetValue()
+    if self.num_images_type.all.GetValue():
```
[the remark this answers is not in the document]
Comment (dwmoreau) [reply xfel/ui/components/xfel_gui_dialogs.py:1496]: I verified that this evaluates to false when it should by adding a print statement in after the else: to verify this is functioning correctly

File: xfel/command_line/cxi_mpi_submit.py [hunk 9]
```diff
@@ -454,17 +456,15 @@ def run(self, argv = None):
...
       extra_str += " %s" % params.input.target
 
     if hasattr(dispatcher_params, 'input') and hasattr(dispatcher_params.input, 'rungroup') and params.input.rungroup is not None:
       data_str += " input.rungroup=%d" % params.input.rungroup
 
-    command = "%s %s output.output_dir=%s %s %s" % (
-      params.input.dispatcher, data_str, output_dir,
-      logging_str, extra_str
-    )
+    command = f"{params.input.dispatcher} {data_str} {logging_str} {extra_str}"
```
Comment (dwpaley) [inline xfel/command_line/cxi_mpi_submit.py:465]:
```suggestion
    command = f"{params.input.dispatcher} {data_str} {logging_str} {extra_str} "
```

File: xfel/command_line/cxi_mpi_submit.py [hunk 10]
```diff
@@ -454,17 +456,15 @@ def run(self, argv = None):
...
+    if params.output.add_output_dir_option:
+      command += f"output.output_dir={output_dir}"
```
Comment (dwpaley) [inline xfel/command_line/cxi_mpi_submit.py:467]:
```suggestion
      command += f" output.output_dir={output_dir}"
```

File: xfel/ui/components/xfel_gui_dialogs.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dwpaley) [reply xfel/ui/components/xfel_gui_dialogs.py:1496]: oh ok now i see that `all` on this line is the specific choice, not a generic wx thing. thanks

**Pull request merged**